alternatives.md

```diff
@@ -0,0 +1,113 @@
+Usages, etc. But it does not define a common representation of AST because
+the goal of the project is to enable easy access to analysis that is done
+by compilers and language SDKs.
+
+Babelfish provides a common representation for ASTs allowing
+to use its output for static analysis by other tools. Queries over UAST
+structure will still allow querying for usages, etc.
+
+## [srclib](https://srclib.org/)
+
+> srclib makes developer tools like code search and static analyzers better. It supports things like jump to definition, find usages, type inference, and documentation generation.
+
+> srclib handles: package detection, global dependency resolution, type inference, querying the graph of definitions and references in code, versioning using different VCS systems, and semantic blaming.
+
+<!-- TODO -->
+
+## [ctags](http://ctags.sourceforge.net/)
+
+> Ctags generates an index (or tag) file of language objects found in source files that allows these items to be quickly and easily located by a text editor or other utility. A tag signifies a language object for which an index entry is available (or, alternatively, the index entry created for that object).
+
+Both Babelfish and Ctags provide positional information for identifiers,
+classes, directives, etc.
+
+The main difference is that Ctags does not provide any form of AST, while
+Babelfish provides native language AST as well as language-independent UAST.
+
+Also, Babelfish ecosystem allows making more complex queries over AST.
+
+## [ANTLR](https://github.com/antlr/antlr4)
+
+> ANTLR (ANother Tool for Language Recognition) is a powerful parser generator for reading, processing, executing, or translating structured text or binary files. It's widely used to build languages, tools, and frameworks. From a grammar, ANTLR generates a parser that can build parse trees and also generates a listener interface (or visitor) that makes it easy to respond to the recognition of phrases of interest.
+
+ANTLR is a language generator with support for most existing languages
+and data formats. But parsers usually produce parse trees that are usually
+different from native language AST, thus requiring additional processing
+to be used for analysis. The structure of parse tree is also language-dependent.
+
+Babelfish provides a correct native language AST as well as
+language-independent UAST for all supported languages.
+
+## [Tree-sitter](https://github.com/tree-sitter/tree-sitter)
+
+> Tree-sitter is a C library for incremental parsing, intended to be used via bindings to higher-level languages. It can be used to build a concrete syntax tree for a program and efficiently update the syntax tree as the program is edited. This makes it suitable for use in text-editing programs.
+
+Tree-sitter provides a simplified AST to be able to execute a limited set
+of queries. It usually cannot provide enough details suitable for static
+analysis. Instead, it focuses on performance and implements real-time
+tree diffing.
```

Comment:
@maxbrunsfeld would know better, but it seems that tree-sitter is an awesome project that is designed to be used in editors, so is optimized for the incremental parsing and low latencies, both of which are not the problems we have in large-scale parsing scenario that bblfsh is addressing.

Tree-sitter also requires maintaining custom grammar definitions for each supported language, that bblfsh does not.

Reply:
+1 on @bzz comment. Tree-Sitter is a very impressive effort but has the downside of creating/maintaining grammars. For its use case, it's perfect. Babelfish is not meant to be used at editor level. For this, I would always recommend tree-sitter.

Reply:
I did not mean to discourage the use of tree-sitter project, sorry if it sounds like this. The project is indeed impressive.

I will update the description to mention custom grammars and all the good features it has.

At the same time, I tried running it, and it seems like an AST that is generated is not a real AST of the language. I wouldn't say it's just a parse tree like in ANTLR, but it missing some details that native AST has.

Reply:
Also is oriented to support partial/malformed code, since is made for editors.

Reply:
> Tree-sitter provides a simplified AST to be able to execute a limited set of queries.

You're right that the output is not an AST, but it's not simplified; it's a Concrete Syntax Tree, so it's a superset of the information contained in an AST. ANTLR produces a CST too, but I don't think ANTLR's trees are designed to use directly for code analysis, whereas Tree-sitter's actually are.

The goal with Tree-sitter grammars is to closely mirror the structure of the AST, so that the resulting tree is *similar* to an AST, but with additional *anonymous* tokens present as well, a little bit like [the ESTree](https://github.com/estree/estree) format for JavaScript syntax trees. [Tree-sitter's APIs](http://tree-sitter.github.io/tree-sitter/using-parsers#named-vs-anonymous-nodes) let you ignore the anonymous tokens if you want.

That said, there is still a big difference which may be important for your use case: Tree-sitter's syntax nodes currently just contain an ordered list of children. The syntax nodes don't have a name for each child like `predicate`, `body`, etc.

Reply:
@maxbrunsfeld Thanks for the clarification! I will update the document accordingly.